### PR TITLE
Adds longest prefix matching for custom indexes.

### PR DESCRIPTION
**Pull request by slackpad**
This adds support for longest prefix matching as needed by https://github.com/hashicorp/consul/pull/1764 (it's not yet integrated over there).

One unfortunate thing is that the null suffixes we add with the `StringFieldIndex` don't work with the algorithm as its currently written - it would need to see if there's a null edge and use that in order to terminate properly. The `LongestPrefix` algorithm is down in the immutable radix tree, so there's no super clean way to get it to understand this.

For my application, I need a custom indexer, similar to the one in the unit test here that can index an empty string, and that will only get added for query templates, so this isn't a big deal. I commented `LongestPrefix` with details about this limitation, and I also made it look for common misconfigurations that will make it not work as expected.

I think we could make under-the-hood improvements to remove this limitation with the `StringFieldIndex` in the future, but this interface would still be useful (it's the `LongestPrefix` form of `First`).

### CODE REVIEW

File: txn_test.go [hunk 1]
```diff
@@ -585,6 +589,187 @@ func TestTxn_InsertGet_Prefix(t *testing.T) {
...
+		if raw != obj1 {
+			t.Fatalf("bad: %#v", raw)
+		}
+
+		raw, err = txn.LongestPrefix("main", "foo_prefix", "foozip")
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+		if raw != obj1 {
+			t.Fatalf("should be nil")
```
Comment (ryanuber) [inline txn_test.go]: But should it? :)

File: txn_test.go [hunk 2]
```diff
@@ -585,6 +589,187 @@ func TestTxn_InsertGet_Prefix(t *testing.T) {
...
+		}
+
+		raw, err = txn.LongestPrefix("main", "foo_prefix", "foozipza")
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+		if raw != obj1 {
+			t.Fatalf("should be nil")
```
Comment (ryanuber) [inline txn_test.go]: Same as above - confusing message

File: txn_test.go (hunk 1 above)
Comment (slackpad) [inline txn_test.go]: aw - cut n' paste - sorry!

File: txn.go [hunk 3]
```diff
@@ -334,6 +334,39 @@ func (txn *Txn) First(table, index string, args ...interface{}) (interface{}, er
 	return value, nil
 }
 
+// LongestPrefix is used to fetch the longest prefix match for the given
+// constraints on the index. Note that this will not work with the memdb
+// StringFieldIndex because it adds null terminators which prevent the
```
Comment (ryanuber) [inline txn.go]: Would it be possible to make a call to `PrefixFromArgs()` to strip the `\x00`? Since this only supports prefix indexers, seems like that should be possible.

**Comment by ryanuber** [comment]
@slackpad Left some comments and a question. Code looks great, just curious about the StringFieldIndex. See the question inline.

File: txn.go (hunk 3 above)
Comment (slackpad) [inline txn.go]: @ryanuber I thought that, too. You can do that to strip the `\x00` off the prefix, but when the `LongestPrefix` routine looks in the tree it hits the node above the actual entry in the index, so it doesn't report anything since that node isn't a leaf. Here's an example:

Say you have `foo` in your tree with a string field indexer. That'll be in there as `foo\x00`. When you do this algorithm for `foobar` it'll scan for the common prefix `foo` just fine and end up at a node with a prefix of `foo` and an edge to `\x00` which is the proper leaf. It doesn't know that it should follow that edge, though, so it just says that there's no entry in there. Because we always put a suffix in there for the string indexer it always runs into this. If you knew to look up `foo\x00` it would work, but you don't know what the longest prefix is when you do a query.

File: txn.go [hunk 4]
```diff
@@ -334,6 +334,39 @@ func (txn *Txn) First(table, index string, args ...interface{}) (interface{}, er
...
+// algorithm from correctly finding a match (it will get to right before the
+// null and fail to find a leaf node). This should only be used where the prefix
+// given is capable of matching indexed entries directly, which typically only
+// applies to a custom indexer. See the unit test for an example.
+func (txn *Txn) LongestPrefix(table, index string, args ...interface{}) (interface{}, error) {
+	// Enforce that this only works on prefix indexes.
+	if !strings.HasSuffix(index, "_prefix") {
+		return nil, fmt.Errorf("index '%s' does not support prefix lookups", index)
```
Comment (dadgar) [inline txn.go]: Not sure I would say does not support prefix lookups because they could pass an index that supports it but didn't append "_prefix". Maybe just "must use prefix lookup on index"

**Comment by dadgar** [comment]
LGTM

File: txn.go (hunk 4 above)
Comment (slackpad) [inline txn.go]: Good call - I'll change that.

File: txn.go (hunk 3 above)
Comment (slackpad) [inline txn.go]: I think the right fix would be to stop adding the `\x00` for the string indexer and instead put that into the compound indexer, because that's where it really matters. That's a bigger change though, and doing that wouldn't affect this interface, just the cautionary comment :-)
